ibis/expr/types/core.py

```diff
@@ -237,6 +238,59 @@ def _find_backend(self) -> BaseBackend:
 
         return backends[0]
 
+    def to_pyarrow_batches(
```

Comment:
Could we generalize this API to make the output formats pluggable? I also like the verb `execute` because it indicates that we are actually executing the ibis expression.

I have something like the following in mind:

```python
expr.execute(format=pd.DataFrame, **options)
expr.execute(pa.Table, **options)
expr.execute(pa.RecordBatch, **options)
expr.execute(polars.DataFrame, **options)
```

We could have string aliases for different formats too.

Reply:
Perhaps we could have an intermediate abstraction, like a `ResultFormatter` or `Executor` which can be overloaded:

```py
@executor.register(PandasBackend, pa.Table)
def pandas_to_pyarrow_table(backend, typ):
    ....


@executor.register(PandasBackend, pa.RecordBatch)
def pandas_to_pyarrow_recordbatch(backend, typ):
    ....


@executor.register(PandasBackend, list)
def pandas_to_plain_python_list(backend, typ):
    ....

@executor.register(PyarrowBackend, pd.DataFrame)
def pyarrow_to_pandas_dataframe(backend, typ):
    ...
```

Reply:
I may not mentioned my primary concern previously. Let's say we have something like the following in the future:

```python
ds = ibis.dataset("s3://bucket/...", schema=...)
expr = ds.filter(...).to_arrow()
```

I would like to execute it using datafusion not pyarrow, how could I define it?
Or I'd like to try out clickhouse and compare the performance of the two?

As far as I can see we need to define both the backend and the format somehow so the user can easily switch engines under the hood.

Reply:
We usually allow to `ibis.<backend>.execute(expr)`. I think that `.to_arrow()` or `.to_csv()` shouldn't trigger an actual execution until `.execute()` is called. 

```python
expr = table.filter(...).to_csv('local.csv')
ibis.pandas.execute(expr)
ibis.pyarrow.execute(expr)
```

I'd also like to avoid hard dependencies like pyarrow or pandas (which is currently). If we hardcode these methods to the internal types then we need to handle multiple optional imports and it can grow in the future.

Reply:
Perhaps we could define specific operation nodes for different formats. 3rdparty backends could define their own operations and the corresponding execution logic too.

Reply:
```py
class Format(Node):
    pass

class PandasFormat(Format):
    options...

class ArrowFormat(Format):
    options...

# it would be still an ibis expression
arrow_expr = t.filter(...).to_arrow(**options)
pandas_expr = t.filter(...).to_pandas(**options)

# this would execute the expression using the datafusion backend
ibis.datafusion.execute(arrow_expr) -> pa.Table
ibis.datafusion.execute(pandas_expr) -> pd.DataFrame
```

Reply:
For the record, I'm not saying that `pyarrow` should be a required dependency, these `to_*` methods can raise a nice exception if their dependencies aren't installed.

Reply:
I agree preserving unbound table behavior is paramount

@cpcloud thoughts here

Reply:
Guarding the imports to avoid the required dependency is pretty straightforward (I'm just working on clearing out a strange issue in the clickhouse tests).

We don't want to hurt the ergonomics of someone using a bound-table, but we also don't want to kneecap unbound tables.

When we have something like `Sinks` in a more concrete form, then we should have no issue executing various file and stream exports on unbound tables via `execute` or with a method on the connection.

Reply:
I'm fine with doing this incrementally, though if we reserve the `Expr.to_pyarrow()` method for actual execution in the upcoming release then we cannot really change it to produce an expression later on.

Reply:
I don't love the idea of an expression that indicates an operation that's really handled in `ibis` itself rather than the backend. A `ToPyarrow` expression couldn't be composed with anything else - it's just a delayed task to eventually execute with `.execute()`. Further, it changes the output of `.execute()`, which currently expects to return a scalar or a pandas thing - this wouldn't play well with our existing `__repr__` code for interactive mode, and makes it harder to tell what `.execute()` will actually return.

Since `to_pyarrow` and `to_pyarrow_batches` are methods on the backend, what's wrong with using `con.to_pyarrow(expr)` to work with unbound tables?

Alternatively (I like this less), what if `to_*` methods execute immediately, but support an optional `execute=False` kwarg to return a delayed node that can later be passed to `con.execute(...)`?

Reply:
> Since to_pyarrow and to_pyarrow_batches are methods on the backend, what's wrong with using con.to_pyarrow(expr) to work with unbound tables?

That's fine, but it is also exposed on `Expr`.

Reply:
> That's fine, but it is also exposed on Expr.

I don't follow? So is `.execute()`, which also doesn't work for unbound tables. All our docs point users to using bound tables - a method version for these conversions is more ergonomic and is what I would expect as a user. For expert users making use of unbound tables, the method on the `Backend` is still available. The existence of one doesn't prevent the user of the other.

Reply:
Unbound tables are an important use case. When using ibis as a library, you often are just merrily constructing expressions without a backend. Eventually you apply that to the backend. To belabor the point: this scenario is most common when ibis is being used as a library not as a tool for interactive data analysis.

We should take unbound tables into account when designing APIs (I am definitely guilty not doing this all the time!) and not treat them as uncommon or not the norm.

That's the library case. On the other hand when you're in interactive mode, you want to type less and have things "just run" or "just work" or "Just Do What I Think This Should".

There are few things to disentangle here, but I think before we do any more work on this we should come up with a few scenarios and design the API from that.

I think the dependencies issue can be resolved :)

The most important thing in my mind is to get those scenarios (let's say 2 or 3).

I'll start a github discussion so we can map those out.

Reply:
Sounds good, until that we can merge this marked as an experimental API.

Reply:
umm how about we have this discussion first before merging